### PR TITLE
grpc proxy support

### PR DESCRIPTION
**Pull request by nic-chen**

### CODE REVIEW

**Comment by membphis** [comment]
nice feature

**Comment by membphis** [comment]
here is the flame graph of current version: [v001.svg.zip](https://github.com/iresty/apisix/files/3513713/v001.svg.zip)

File: lua/apisix.lua [hunk 1]
```diff
@@ -159,6 +159,11 @@ function _M.http_access_phase()
         return core.response.exit(404)
     end
 
+    --
+    if route.value.service_protocol=="grpc" then
```
Comment (moonming) [inline lua/apisix.lua]: code style:  need one space on both sides of `==`

File: lua/apisix.lua [hunk 2]
```diff
@@ -197,11 +202,74 @@ function _M.http_access_phase()
     run_plugin("access", plugins, api_ctx)
 end
 
+function _M.grpc_access_phase()
+    local ngx_ctx = ngx.ctx
+    local api_ctx = ngx_ctx.api_ctx
+
+    if api_ctx == nil then
```
Comment (moonming) [inline lua/apisix.lua]: code style: if not api_ctx then

File: t/debug-mode.t [hunk 3]
```diff
@@ -46,6 +46,7 @@ loaded plugin and sort by priority: 1003 name: limit-conn
 loaded plugin and sort by priority: 1002 name: limit-count
 loaded plugin and sort by priority: 1001 name: limit-req
 loaded plugin and sort by priority: 1000 name: node-status
+loaded plugin and sort by priority: 506 name: grpc-proxy
```
Comment (moonming) [inline t/debug-mode.t]: need to add test cases for http -> grpc.

File: lua/apisix/plugins/grpc-proxy/util.lua [hunk 4]
```diff
@@ -0,0 +1,82 @@
+local pb = require("pb")
+local json
+if not os.getenv("LUAUNIT") then
```
Comment (moonming) [inline lua/apisix/plugins/grpc-proxy/util.lua]: useless code?

File: lua/apisix/plugins/grpc-proxy/util.lua [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    json = require("cjson")
+end
+
+local _M = {}
+
+_M.file_exists = function(file)
```
Comment (moonming) [inline lua/apisix/plugins/grpc-proxy/util.lua]: useless function?

File: lua/apisix/plugins/grpc-proxy/util.lua [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            sub, err = _M.map_message(field_type, default_values)
+            if err then
+                return nil, err
+            end
+            request[name] = sub
+        else
+            request[name] = get_from_request(name, field_type) or default_values[name] or nil
+        end
+    end
+    return request, nil
```
Comment (moonming) [inline lua/apisix/plugins/grpc-proxy/util.lua]: code style:
```
return request
```

File: lua/apisix/plugins/grpc-proxy/proto.lua [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+    return _p.loaded
+end
+
+
+local _M = {version = 0.1}
+
+
+function _M.new(proto_id)
+    local key = "/proto"..proto_id
```
Comment (moonming) [inline lua/apisix/plugins/grpc-proxy/proto.lua]: code style

**Comment by membphis** [comment]
have problem:

image

File: lua/apisix/core/schema.lua [hunk 8]
```diff
@@ -257,6 +258,10 @@ _M.route = [[{
             },
             "uniqueItems": true
         },
+        "service_protocol": {
```
Comment (moonming) [inline lua/apisix/core/schema.lua]: the value only be `grpc` now, so we should limit it.

Comment (moonming) [reply lua/apisix/core/schema.lua]: and need to add test case for that.

File: rockspec/apisix-dev-1.0-0.rockspec [hunk 9]
```diff
@@ -28,6 +28,7 @@ dependencies = {
     "opentracing-openresty = 0.1",
     "lua-resty-radixtree = 0.4",
     "lua-resty-iputils = 0.3.0-1",
+    "lua-protobuf = 0.3.1",
```
Comment (moonming) [inline rockspec/apisix-dev-1.0-0.rockspec]: need add copyright

File: t/plugin/grpc-proxy.t [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+=== TEST 3: hit route
+--- request
+GET /grpctest
+--- response_body eval
+qr/\{"message":"Hello "\}/
+--- no_error_log
+[error]
+
+
+=== TEST 2: wrong service protocol
```
Comment (moonming) [inline t/plugin/grpc-proxy.t]: wrong index of test

**Pull request merged**